src-java/ping-topology/ping-storm-topology/src/test/java/org/openkilda/wfm/topology/ping/PingTopologyTest.java

```diff
@@ -373,31 +390,164 @@ public void flowFetcherHaFlowOneSubFlowWithEndpointEqualsYPoint() {
+                        ImmutableMap.of("ha_flow_id", HA_FLOW_ID_1,
+                                "flowid", SUB_FLOW_ID_A,
+                                "direction", "reverse",
+                                "status", "success"), 1),
+                new Datapoint("kilda.flow.latency", 0L,
+                        ImmutableMap.of("ha_flow_id", HA_FLOW_ID_1,
+                                "flowid", SUB_FLOW_ID_B,
+                                "direction", "forward",
+                                "status", "success"), 1),
+                new Datapoint("kilda.flow.latency", 0L,
+                        ImmutableMap.of("ha_flow_id", HA_FLOW_ID_1,
+                                "flowid", SUB_FLOW_ID_B,
+                                "direction", "reverse",
+                                "status", "success"), 1)
+        );
+
+        createCompleteHaFlowWithPeriodicPing();
+
+        List<CommandMessage> pingCommands = speakerConsumer.assertNAndPoll(4, 2000, CommandMessage.class);
+        pingCommands.forEach(this::sendSpeakerAnswer);
+
+        List<Datapoint> resultDataPoints = otsdbConsumer.assertNAndPoll(4, Datapoint.class);
+        assertThat(resultDataPoints, Matchers.containsInAnyOrder(expectedDatapoints.toArray()));
+
+        haFlowRepository.remove(HA_FLOW_ID_1);
+        disableHaFlowAndAssert();
+    }
+
+
+    @Test
+    public void flowFetcherHaPeriodicPingOneFails() {
+        final List<Datapoint> expectedDatapoints = Lists.newArrayList(
+                new Datapoint("kilda.flow.latency", 0L,
+                        ImmutableMap.of("ha_flow_id", HA_FLOW_ID_1,
+                                "flowid", SUB_FLOW_ID_A,
+                                "direction", "forward",
+                                "status", "success"), 1),
+                new Datapoint("kilda.flow.latency", 0L,
+                        ImmutableMap.of("ha_flow_id", HA_FLOW_ID_1,
+                                "flowid", SUB_FLOW_ID_A,
+                                "direction", "reverse",
+                                "status", "error"), -1),
+                new Datapoint("kilda.flow.latency", 0L,
+                        ImmutableMap.of("ha_flow_id", HA_FLOW_ID_1,
+                                "flowid", SUB_FLOW_ID_B,
+                                "direction", "forward",
+                                "status", "success"), 1),
+                new Datapoint("kilda.flow.latency", 0L,
+                        ImmutableMap.of("ha_flow_id", HA_FLOW_ID_1,
+                                "flowid", SUB_FLOW_ID_B,
+                                "direction", "reverse",
+                                "status", "error"), -1)
+        );
+
+        createCompleteHaFlowWithPeriodicPing();
+
+        sendHaFlowPeriodicPingCommand(HA_FLOW_ID_1);
+        List<CommandMessage> pingCommands = speakerConsumer.assertNAndPoll(4, 1000, CommandMessage.class);
+
+        // send only 2 answers
```

Comment:
a funny fact: just two lines like 503 sending 2 answers would be shorter than a loop with a comment

Reply:
done